tests/ui/consts/const-mut-refs-crate.rs

```diff
@@ -0,0 +1,37 @@
+//@ run-pass
+//@ aux-build:const_mut_refs_crate.rs
+
+#![feature(const_mut_refs)]
+
+//! Regression test for https://github.com/rust-lang/rust/issues/79738
+//! Show how we are not duplicating allocations anymore. Statics that
+//! copy their value from another static used to also duplicate
+//! memory behind references.
+
+extern crate const_mut_refs_crate as other;
+
+use other::{
+    inner::{INNER_MOD_BAR, INNER_MOD_FOO},
+    BAR, FOO,
+};
+
+pub static LOCAL_FOO: &'static i32 = &41;
+pub static LOCAL_BAR: &'static i32 = LOCAL_FOO;
+pub static mut COPY_OF_REMOTE_FOO: &'static mut i32 = unsafe { FOO };
+
+static DOUBLE_REF: &&i32 = &&99;
+static ONE_STEP_ABOVE: &i32 = *DOUBLE_REF;
+static mut DOUBLE_REF_MUT: &mut &mut i32 = &mut &mut 99;
+static mut ONE_STEP_ABOVE_MUT: &mut i32 = unsafe { *DOUBLE_REF_MUT };
+
+pub fn main() {
+    unsafe {
+        assert_eq!(FOO as *const i32, BAR as *const i32);
+        assert_eq!(INNER_MOD_FOO as *const i32, INNER_MOD_BAR as *const i32);
```

Comment:
Looks like this case still depends on deduplication of anonymous constants, causing the test to fail with cg_clif.

Reply:
yea it's more like a canary test than anything we want to necessarily guarantee. Though maybe we could start guaranteeing the deduplication by generating canonical static items whose id/name is their contents' stable hash?

Reply:
> yea it's more like a canary test than anything we want to necessarily guarantee.

Would it be possible to split the test into a guaranteed half and a canary half? That way the guaranteed half can be tested with cg_clif and the canary half can be ignored.

Reply:
yea that sounds reasonable. Will open a PR shortly

Reply:
At the MIR level, `INNER_MOD_FOO` has `43` in a promoted, while `INNER_MOD_BAR` has it in an anonymous allocation:

```rust
static INNER_MOD_FOO: &i32 = {
    let mut _0: &i32;
    let _1: &i32;
    let _2: i32;
    let mut _3: &i32;

    bb0: {
        StorageLive(_1);
        _3 = const INNER_MOD_FOO::promoted[0];
        _1 = &(*_3);
        _0 = &(*_1);
        StorageDead(_1);
        return;
    }
}

const INNER_MOD_FOO::promoted[0]: &i32 = {
    let mut _0: &i32;
    let mut _1: i32;

    bb0: {
        _1 = const 43_i32;
        _0 = &_1;
        return;
    }
}

static INNER_MOD_BAR: &i32 = {
    let mut _0: &i32;
    let mut _1: &&i32;

    bb0: {
        StorageLive(_1);
        _1 = const {alloc1: &&i32};
        _0 = (*_1);
        StorageDead(_1);
        return;
    }
}

alloc1 (static: INNER_MOD_FOO, size: 8, align: 8) {
    ╾─────alloc3<imm>─────╼                         │ ╾──────╼
}

alloc3 (size: 4, align: 4) {
    2b 00 00 00                                     │ +...
}
```

Reply:
oh right, this is using promotion 🤦 Does the test work for you if you change the body of `INNER_MOD_FOO` to

```rust
{
    let x = 43;
    &{x}
}
```

Reply:
That will test that specific assertion. For the rest equivalent changes are necessary.

Reply:
Some tests maybe also should be swapped to mutable references. The test is called "const-mut-refs" after all. ;)

Reply:
I created https://github.com/rust-lang/rust/pull/122719 to ensure we get sensible ICEs from now on